core-client/src/lib.rs

```diff
@@ -202,6 +204,80 @@ where
 	}
 }
 
+/// Client for raw JSON RPC requests
+#[derive(Clone)]
+pub struct RawClient(RpcChannel);
+
+impl From<RpcChannel> for RawClient {
+	fn from(channel: RpcChannel) -> Self {
+		RawClient(channel)
+	}
+}
+
+impl RawClient {
+	/// Call RPC with raw JSON
+	pub fn call_method(&self, method: &str, params: Params) -> impl Future<Item=Value, Error=RpcError> {
+		let (sender, receiver) = oneshot::channel();
+		let msg = RpcMessage {
+			method: method.into(),
+			params,
+			sender,
+		};
+		self.0
+			.to_owned()
+			.send(msg)
+			.map_err(|error| RpcError::Other(error.into()))
+			.and_then(|_| RpcFuture::new(receiver))
+	}
+}
+
+/// Client for typed JSON RPC requests
+#[derive(Clone)]
+pub struct TypedClient(RawClient);
+
+impl From<RpcChannel> for TypedClient {
+	fn from(channel: RpcChannel) -> Self {
+		TypedClient(channel.into())
+	}
+}
+
+impl TypedClient {
+	/// Create new TypedClient
+	pub fn new(raw_cli: RawClient) -> Self {
+		TypedClient(raw_cli)
+	}
+
+	/// Call RPC with serialization of request and deserialization of response
+	pub fn call_method<T: Serialize, R: DeserializeOwned + 'static>(
+		&self,
+		method: &str,
+		returns: &'static str,
+		args: T,
+	) -> impl Future<Item=R, Error=RpcError> {
+		let args = serde_json::to_value(args)
+			.expect("Only types with infallible serialisation can be used for JSON-RPC");
+		let params = match args {
+			Value::Array(vec) => Params::Array(vec),
+			Value::Null => Params::None,
+			_ => panic!("RPC params should serialize to a JSON array, or null"),
```

Comment:
Wasn't sure what to do here. This was not possible when this was in `derive` because the params are guaranteed to be a Tuple.

Reply:
Maybe just return an `RpcError`?

Reply:
Yes I did that originally before changing to panic, I think that might be better after all.